File: 0820-find-eventual-safe-states/0820-find-eventual-safe-states.java
class Solution {
    public boolean dfs (int node, int [] visited, int [] [] graph) {
        visited [node] = 2;

        for (int it : graph [node]) {
            if (visited [it] == 0) {
                if (dfs (it, visited, graph)) {
                    return true;
                }
            } else if (visited [it] == 2) {
                return true;
            }
        }

        visited [node] = 1;
        return false;
    }
    public List <Integer> eventualSafeNodes (int [] [] graph) {
        int V = graph.length;
        int [] visited = new int [V];

        for (int i = 0; i < V; i++) {
            if (visited [i] == 0) {
                dfs (i, visited, graph);
            }
        }

        List <Integer> safeNodes = new ArrayList <> ();
        for (int i = 0; i < V; i++) {
            if (visited [i] == 1) {
                safeNodes.add (i);
            }
        }
        return safeNodes;
    }
}

File: 0820-find-eventual-safe-states/0820-find-eventual-safe-states.c
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */

bool DFS (int **graph, int *graphColSize, int *safe, int *visited, int node) {
    bool isSafe = true;

    for (int i = 0; i < graphColSize [node]; i++) {
        int neighbour = graph [node] [i];
        if (!safe [neighbour] || visited [neighbour]) {
            isSafe = false;
            break;
        }
        if (safe [neighbour] == 1) {
            continue;
        }
        visited [neighbour] = 1;
        if (DFS (graph, graphColSize, safe, visited, neighbour) == false) {
            isSafe = false;
            break;
        }
        visited [neighbour] = 0;
    }
    safe [node] = isSafe ? 1 : 0;
    return isSafe;
}

int* eventualSafeNodes (int** graph, int graphSize, int* graphColSize, int* returnSize) {
    int* safe = malloc (graphSize * sizeof (int));
    int* visited = malloc (graphSize * sizeof (int));

    for (int i = 0; i < graphSize;i++) {
        safe [i] = -1;
        visited [i] = 0;
    }

    for (int node = 0; node < graphSize; node++)
        if (safe [node] == -1) {
            visited [node] = 1;
            DFS (graph, graphColSize, safe, visited, node);
            visited [node] = 0;
        }

    int *returnArray = malloc (graphSize * sizeof (int));
    *returnSize = 0;
    for (int node = 0; node < graphSize; node++)
        if (safe [node] == 1)
            returnArray [(*returnSize)++] = node;
    
    return returnArray;
}